File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/forth/forth.js
//
// This is only a SKELETON file for the 'Forth' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Forth {
  constructor() {
    this._stack = [];
    this._dictionary = new Map();
    // Initialize built-in words
    this._dictionary.set('+', this.add.bind(this));
    this._dictionary.set('-', this.subtract.bind(this));
    this._dictionary.set('*', this.multiply.bind(this));
    this._dictionary.set('/', this.divide.bind(this));
    this._dictionary.set('dup', this.dup.bind(this));
    this._dictionary.set('drop', this.drop.bind(this));
    this._dictionary.set('swap', this.swap.bind(this));
    this._dictionary.set('over', this.over.bind(this));
  }

  add() {
    if (this._stack.length < 2) throw new Error('Stack empty');
    const b = this._stack.pop();
    const a = this._stack.pop();
    this._stack.push(a + b);
  }

  subtract() {
    if (this._stack.length < 2) throw new Error('Stack empty');
    const b = this._stack.pop();
    const a = this._stack.pop();
    this._stack.push(a - b);
  }

  multiply() {
    if (this._stack.length < 2) throw new Error('Stack empty');
    const b = this._stack.pop();
    const a = this._stack.pop();
    this._stack.push(a * b);
  }

  divide() {
    if (this._stack.length < 2) throw new Error('Stack empty');
    const b = this._stack.pop();
    const a = this._stack.pop();
    if (b === 0) throw new Error('Division by zero');
    this._stack.push(Math.floor(a / b));
  }

  dup() {
    if (this._stack.length < 1) throw new Error('Stack empty');
    this._stack.push(this._stack[this._stack.length - 1]);
  }

  drop() {
    if (this._stack.length < 1) throw new Error('Stack empty');
    this._stack.pop();
  }

  swap() {
    if (this._stack.length < 2) throw new Error('Stack empty');
    const b = this._stack.pop();
    const a = this._stack.pop();
    this._stack.push(b);
    this._stack.push(a);
  }

  over() {
    if (this._stack.length < 2) throw new Error('Stack empty');
    this._stack.push(this._stack[this._stack.length - 2]);
  }

  tokenize(input) {
    return input.split(/\s+/).filter(t => t !== '').map(t => {
      if (/^-?\d+$/.test(t)) {
        return parseInt(t, 10);
      }
      return t.toLowerCase();
    });
  }

  evaluate(input) {
    const tokens = this.tokenize(input);
    const tokenStack = [{ tokens, index: 0 }];
    const tempDefinitions = new Map();
    
    while (tokenStack.length > 0) {
      const currentFrame = tokenStack[tokenStack.length - 1];
      const { tokens } = currentFrame;
      
      if (currentFrame.index >= tokens.length) {
        tokenStack.pop();
        continue;
      }
      
      const token = tokens[currentFrame.index++];
      
      if (token === ':') {
        if (currentFrame.index >= tokens.length) {
          throw new Error('Invalid definition');
        }
        const wordName = tokens[currentFrame.index++];
        if (typeof wordName !== 'string' || /^-?\d+$/.test(wordName)) {
          throw new Error('Invalid definition');
        }
        
        let j = currentFrame.index;
        while (j < tokens.length && tokens[j] !== ';') {
          j++;
        }
        if (j >= tokens.length) {
          throw new Error('Definition not terminated');
        }
        
        const definitionTokens = tokens.slice(currentFrame.index, j);
        const name = wordName.toLowerCase();
        this._dictionary.set(name, definitionTokens);
        currentFrame.index = j + 1;
      } else {
        if (typeof token === 'number') {
          this._stack.push(token);
        } else {
          const word = token.toLowerCase();
          const definition = this._dictionary.get(word);
          if (!definition) {
            throw new Error('Unknown command');
          }
          if (typeof definition === 'function') {
            definition.call(this);
          } else {
            tokenStack.push({ tokens: definition, index: 0 });
          }
        }
      }
    }
  }

  get stack() {
    return [...this._stack];
  }
}

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/forth/forth.go
package forth

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

func Forth(input []string) ([]int, error) {
	stack := []int{}
	customWords := make(map[string][]string)

	for _, line := range input {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		if strings.HasPrefix(line, ":") {
			parts := strings.Fields(line)
			if len(parts) < 4 || parts[0] != ":" || parts[len(parts)-1] != ";" {
				return nil, errors.New("invalid definition")
			}
			wordName := strings.ToLower(parts[1])
			if _, err := strconv.Atoi(wordName); err == nil {
				return nil, errors.New("word name cannot be a number")
			}
			definition := parts[2 : len(parts)-1]
			for _, token := range definition {
				if strings.ToLower(token) == wordName {
					return nil, errors.New("recursive definition")
				}
			}
			customWords[wordName] = definition
		} else {
			tokens := strings.Fields(line)
			processingQueue := tokens

			for len(processingQueue) > 0 {
				token := processingQueue[0]
				processingQueue = processingQueue[1:]

				tokenLower := strings.ToLower(token)
				if num, err := strconv.Atoi(token); err == nil {
					stack = append(stack, num)
					continue
				}

				if def, ok := customWords[tokenLower]; ok {
					processingQueue = append(def, processingQueue...)
					continue
				}

				switch tokenLower {
				case "+":
					if len(stack) < 2 {
						return nil, errors.New("stack underflow")
					}
					a, b := stack[len(stack)-1], stack[len(stack)-2]
					stack = stack[:len(stack)-2]
					stack = append(stack, b+a)
				case "-":
					if len(stack) < 2 {
						return nil, errors.New("stack underflow")
					}
					a, b := stack[len(stack)-1], stack[len(stack)-2]
					stack = stack[:len(stack)-2]
					stack = append(stack, b-a)
				case "*":
					if len(stack) < 2 {
						return nil, errors.New("stack underflow")
					}
					a, b := stack[len(stack)-1], stack[len(stack)-2]
					stack = stack[:len(stack)-2]
					stack = append(stack, b*a)
				case "/":
					if len(stack) < 2 {
						return nil, errors.New("stack underflow")
					}
					a := stack[len(stack)-1]
					if a == 0 {
						return nil, errors.New("division by zero")
					}
					b := stack[len(stack)-2]
					stack = stack[:len(stack)-2]
					stack = append(stack, b/a)
				case "dup":
					if len(stack) < 1 {
						return nil, errors.New("stack underflow")
					}
					stack = append(stack, stack[len(stack)-1])
				case "drop":
					if len(stack) < 1 {
						return nil, errors.New("stack underflow")
					}
					stack = stack[:len(stack)-1]
				case "swap":
					if len(stack) < 2 {
						return nil, errors.New("stack underflow")
					}
					stack[len(stack)-1], stack[len(stack)-2] = stack[len(stack)-2], stack[len(stack)-1]
				case "over":
					if len(stack) < 2 {
						return nil, errors.New("stack underflow")
					}
					stack = append(stack, stack[len(stack)-2])
				default:
					return nil, fmt.Errorf("undefined word: %s", token)
				}
			}
		}
	}

	return stack, nil
}
